dace/codegen/tools/type_inference.py

```diff
@@ -0,0 +1,580 @@
+"""
+    Type inference: traverses code and returns types for all undefined symbols according to C semantics
+    infer() has a lenient implementation: if something it not inferred (for example an unsupported construct) it will not
+        return anything and it will not produce errors
```

Comment:
Are we sure we don't want to fail if something cannot be inferred? For OpenCL we need to fail in this case. What will the dictionary contain in this case, `None`?

Reply:
Probably it will contain `None`.

In `cppunparse` we fail if we try to unparse a statement that is not expressible in C++.
Don't  know if it is ok to fail also here, because then we will  tailor type inference on C++

Reply:
Do you mean C++ or C? If it's not expressible in C, we want to fail I think?

Reply:
I think it's fine to return `None`s for un-inferrable variables. Other interface users may replace that with e.g., default types or return a notification to the user.